logic/Rule.java

```diff
@@ -402,12 +402,16 @@ public void validate(LogicManager logicMgr, ConceptManager conceptMgr) {
         }
 
         private void validateInsertable(LogicManager logicMgr) {
-            FunctionalIterator<Map<Identifier.Variable.Name, Label>> whenCombinations = logicMgr.typeInference().typePermutations(rule.when, false);
-            Set<Map<Identifier.Variable.Name, Label>> allowedThenCombinations = logicMgr.typeInference().typePermutations(rule.then, true).toSet();
+            Set<Identifier.Variable.Retrievable> sharedIDs = iterate(rule.then.identifiers())
+                    .filter(thenID -> thenID.isName() && rule.when.identifiers().contains(thenID))
+                    .map(Identifier.Variable::asRetrievable).toSet();
+            FunctionalIterator<Map<Identifier.Variable.Name, Label>> whenCombinations = logicMgr.typeInference().typePermutations(rule.when, false, sharedIDs);
+            Set<Map<Identifier.Variable.Name, Label>> insertableThenCombinations = logicMgr.typeInference().typePermutations(rule.then, true, sharedIDs).toSet();
```

Comment:
permutations for when and then only return permutations of the shared variables - a significantly smaller number for each!

Reply:
Also, why can't we combine the when and then into a single conjunction and do type inference over that?